### DISCUSSION TITLE
Change over to SLF4j

### DISCUSSION THREAD

**Discussion by MikeNeilson** [opendcs/opendcs#487] [Development Practices]
@opendcs/opendcs-core-devs @opendcs/opendcs-triage @opendcs/opendcs-dcpmon-devs

I've just merged in the branch that starts the migration to using SLF4j for the logging fascade. If you are writing new code, please use it, if you are adjusting log statements, please use it for the new code. If you have time in a file you're working it can make sense to just switch it over.

The backend continues to be the original OpenDCS custom logger. We will change that in the future, but for now just the fascade gets us an improvement.

**Comment by ktarbet** [comment]
Trying out using SLF4j -- I came across this code :
```
 String msg = "Can't instantiate object of type '"
                + className + "': " + ex;
            System.err.println(msg);
            log.error(msg,ex);
```
We could change to 
```
log.error("Can't instantiate object of type '{}'",className,ex);
```
And drop the System.err.println  , which feels redudant?
Or should we keep System.err for consistency?

**Reply by MikeNeilson** [reply]
Drop the system.err.println, actually part of the point of using slf4j to handle those cases easier.
Printing directly to stderr (outside of an actual interactive CLI app) is a bad idea.

**Comment by agilmore2** [comment]
So do we have a mapping from the old log levels to the log.x() methods? 
There doesn't seem to be a debug2 level?

**Reply by MikeNeilson** [reply]
There isn't. SLF4j intentional skips that.

https://github.com/opendcs/opendcs/blob/c408ba2e4976ba8c0854c2f517ed56b354163559/java/opendcs-slf4j-provider/src/main/java/org/opendcs/utils/logging/SLF4JLogger.java#L104

debug2,debug3 -> trace in slf4j

**Comment by agilmore2** [comment]
Thanks. I moved a couple of debug2 statements to log.debug instead, but I
think that makes sense.

**Comment by MikeNeilson** [comment]
Yeah, definitelt  judgement call.

One other thing, if you have log statements in a loop, to like print a
list, wrap it in an `if level enabled` check.

**Comment by MikeNeilson** [comment]
FYI, I'm planning on defaulting to https://logback.qos.ch/ for the actual logger. The XML config is flexible, log filename substitution (e.g. keeping the -l switch) is trivial. We *may* also be able to map -d N to the appropriate levels with minimal effort... but I'm not going to spend too much time on that.

Other justifications:
1. Written by the same people that do SLF4j so all the features map
2. Has built-in Async support (the at least perceived need of brought us to this situtation)
3. Plenty of output options, including JSON which will be create for any cloud hosted systems and log ingestion
4. Good documentation on how to use the fancy features, when to use them, and well defined interfaces such that hooking into/from the GUI shouldn't be too difficult.

Just using java.util.logging was considered, however we would likely need to do more custom work as the current sfl4j logging provider doesn't support the MappedDiagnosticContext (MDC), or Markers. MDC is already using in the routing scheduler so that log messages are tied to the write source. Markers will very likely be used in the Alarm system to replace how that was previously handled.

That all said, please bring on any counter arguments.
